From 7f36fbf774c973ef3e0d2581e68c33cc0da76a3e Mon Sep 17 00:00:00 2001
From: Laurent Pinchart <laurent.pinchart+renesas@ideasonboard.com>
Date: Mon, 17 Jun 2013 10:41:23 +0900
Subject: [PATCH 497/715] drm: rcar-du: Add a module argument to enable
 panning in fbdev emulation

When set to true, the fbdev_pan argument forces the driver to allocate a
frame buffer for fbdev emulation twice as high as the screen size. This
allows implementing page flipping through panning with the emulated
fbdev device.

Signed-off-by: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
(cherry picked from commit 21cb3c3a30ecd995bbb4b1b5a9be314af93993cd)

Signed-off-by: Ryo Kataoka <ryo.kataoka.wt@renesas.com>
---
 drivers/gpu/drm/rcar-du/rcar_du_kms.c |    8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/rcar-du/rcar_du_kms.c b/drivers/gpu/drm/rcar-du/rcar_du_kms.c
index 70f6116..a670211 100644
--- a/drivers/gpu/drm/rcar-du/rcar_du_kms.c
+++ b/drivers/gpu/drm/rcar-du/rcar_du_kms.c
@@ -11,6 +11,8 @@
  * (at your option) any later version.
  */
 
+#include <linux/moduleparam.h>
+
 #include <drm/drmP.h>
 #include <drm/drm_crtc.h>
 #include <drm/drm_crtc_helper.h>
@@ -24,6 +26,9 @@
 #include "rcar_du_regs.h"
 #include "rcar_du_vga.h"
 
+static bool rcar_du_fbdev_pan = false;
+module_param_named(fb_cma_pan, rcar_du_fbdev_pan, bool, 0444);
+
 /* -----------------------------------------------------------------------------
  * Format helpers
  */
@@ -245,7 +250,8 @@ int rcar_du_modeset_init(struct rcar_du_device *rcdu)
 	drm_helper_disable_unused_functions(dev);
 
 	fbdev = drm_fbdev_cma_init(dev, 32, dev->mode_config.num_crtc,
-				   dev->mode_config.num_connector, 1);
+				   dev->mode_config.num_connector,
+				   rcar_du_fbdev_pan ? 2 : 1);
 	if (IS_ERR(fbdev))
 		return PTR_ERR(fbdev);
 
-- 
1.7.10.4

